agent/consul/state/memdb_wrapper.go

```diff
@@ -0,0 +1,92 @@
+package state
+
+import (
+	"github.com/hashicorp/go-memdb"
+)
+
+// memDBWrapper is a thin wrapper around memdb.DB which enables TrackChanges on
+// all write transactions. When the transaction is committed the changes are
+// sent to the eventPublisher which will create and emit change events.
+type memDBWrapper struct {
+	db *memdb.MemDB
+	// TODO: add publisher
+}
+
+// Txn exists to maintain backwards compatibility with memdb.DB.Txn. Preexisting
+// code may use it to create a read-only transaction, but it will panic if called
+// with write=true.
+//
+// Deprecated: use either ReadTxn, or WriteTxn.
+func (db *memDBWrapper) Txn(write bool) *txnWrapper {
+	if write {
+		panic("don't use db.Txn(true), use db.WriteTxn(idx uin64)")
+	}
+	return db.ReadTxn()
+}
+
+// ReadTxn returns a read-only transaction which behaves exactly the same as
+// memdb.Txn
+func (db *memDBWrapper) ReadTxn() *txnWrapper {
+	return &txnWrapper{Txn: db.db.Txn(false)}
+}
+
+// WriteTxn returns a wrapped memdb.Txn suitable for writes to the state store.
+// It will track changes and publish events for the changes when Commit
+// is called.
+//
+// The idx argument must be the index of the current Raft operation. Almost
+// all mutations to state should happen as part of a raft apply so the index of
+// the log being applied can be passed to WriteTxn.
+// The exceptional cases are transactions that are executed on an empty
+// memdb.DB as part of Restore, and those executed by tests where we insert
+// data directly into the DB. These cases may use WriteTxnRestore.
+func (db *memDBWrapper) WriteTxn(idx uint64) *txnWrapper {
+	t := &txnWrapper{
+		Txn:   db.db.Txn(true),
+		Index: idx,
+	}
+	t.Txn.TrackChanges()
+	return t
+}
+
+// WriteTxnRestore returns a wrapped RW transaction that does NOT have change
+// tracking enabled. This should only be used in Restore where we need to
+// replace the entire contents of the Store without a need to track the changes.
+// WriteTxnRestore uses a zero index since the whole restore doesn't really occur
+// at one index - the effect is to write many values that were previously
+// written across many indexes.
+func (db *memDBWrapper) WriteTxnRestore() *txnWrapper {
+	t := &txnWrapper{
+		Txn:   db.db.Txn(true),
+		Index: 0,
+	}
+	return t
+}
+
+// txnWrapper wraps a memdb.Txn to capture changes and send them to the
+// EventPublisher.
+//
+// This can not be done with txn.Defer because the callback passed to Defer is
+// invoked after commit completes, and because the callback can not return an
+// error. Any errors from the callback would be lost,  which would result in a
+// missing change event, even though the state store had changed.
+type txnWrapper struct {
```

Comment:
I'm tempted to call this simply `txn`. I tried to think of another names, but I failed to come up with anything that I felt was an improvement. `Wrapper` here does not tell me a whole lot more about why or when it should be used. I believe everything should be using this transaction type, so something like `txn` may help imply that.

This name is widely used, so it would be great to rename it now if we want to.

Thoughts on naming this `txn` ?

Reply:
I agree overall.

My slight hesitation with `txn` is that we do sometimes use that a variable or argument name too (although I think `tx` is even more common there) and that could be really confusing if the type name got shadowed by a variable name? Effective go encourages not naming pacages or types the same as things that are often used as variable names IIRC and I think this arguable qualifies. We could port all uses of `txn` to `tx` but is there still potential for confusion if it's re-introduced?

I can't think of great alternatives. We could use `storeTxn` which overall stutters (state.stateTxn)/seems redundant but it's not a public export wo will never be used in the qualified form in code. That implies it's the transaction type to use on our state store which is useful while keeping it clearly distinct from any variable name we might choose to store one in. Thoughts?